lib/rspec/core/configuration.rb

```diff
@@ -2033,6 +2032,12 @@ def on_existing_matching_groups(meta)
         end
       end
 
+      def add_hook_to_existing_groups(prepend_or_append, position, *meta, &block)
```

Comment:
I thought the behaviour of helper modules and hooks were too different to use the same logic - helper modules are currently applied to every existing matching group, whereas hooks need to only be applied to the highest level matching groups.

However it looks like when creating a new example group, helper modules work that way too: https://github.com/rspec/rspec-core/blob/v3.5.0/lib/rspec/core/configuration.rb#L2044

I pushed a new commit that updates the behaviour of `on_existing_matching_groups`, and uses it to apply both helper modules and hooks. I'm not happy with some things about the implementation - I'll comment inline where I suspect there might be a better way to do something.

If we want to go in this direction, I think we should fix the behaviour of helper modules separately, and then rebase this change to just fix the issues with hooks afterwards.